core/internal/runhistoryreader/parquet/iterator/row_iterator.go

```diff
@@ -0,0 +1,391 @@
+	return func(c *IteratorConfig) {
+		c.Key = TimestampKey
+		c.Min = page.MinTimestamp
+		c.Max = page.MaxTimestamp
+	}
+}
+
+func NewRowIterator(
+	ctx context.Context,
+	reader *pqarrow.FileReader,
+	keys []string,
+	opt RowIteratorOption, // We require atleast one of the history page range or events page range to be passed in
+) (*Iterator, error) {
+	config := IteratorConfig{
+		Key: StepKey, // TODO: make this configurable
+		Min: float64(0),
+		Max: math.MaxInt64,
+	}
+	config.Apply(opt)
+	if config.Max == 0 {
+		// In case the caller code accidentally passes in a config but only explicitly sets the Min,
+		// in which case Max defaults to 0. In that sencario, if we don't do this, in Next() we'll skip all rows
+		config.Max = math.MaxInt64
+	}
+
+	// Check if this file is empty
+	if reader.ParquetReader().NumRows() == 0 {
+		rgi := &Iterator{
+			current: &NoopRowIterator{},
+		}
+		return rgi, nil
+	}
+
+	if reader.Props.BatchSize <= 0 {
+		if len(keys) > 0 {
+			reader.Props.BatchSize = int64(batchSizeForSchemaSize(len(keys)))
+		} else {
+			numColumns := reader.ParquetReader().MetaData().Schema.NumColumns()
+			reader.Props.BatchSize = int64(batchSizeForSchemaSize(numColumns))
+		}
+	}
+
+	selectedColumns := keys
+	var rowGroupIndices []int
+
+	// if we need pagination, then there's a bit more work to do
+	needPage := config.Min != 0 || config.Max != math.MaxInt64
+	if needPage {
+		// check what row groups are valid
+		var err error
+		rowGroupIndices, err = selectRowGroups(reader.ParquetReader(), config)
+		if err != nil {
+			return nil, err
+		}
+
+		// and include the index key if we're not selecting it
+		if len(keys) > 0 {
+			if !slices.Contains(keys, config.Key) {
+				selectedColumns = append(selectedColumns, config.Key)
+			}
+		}
+	} else {
+		rowGroupIndices = make([]int, reader.ParquetReader().NumRowGroups())
+		for i := range rowGroupIndices {
+			rowGroupIndices[i] = i
+		}
+	}
+
+	columnIndices, err := selectColumns(reader.ParquetReader().MetaData().Schema, selectedColumns)
+	if err != nil {
+		rgi := &Iterator{
+			current: &NoopRowIterator{},
+		}
+		return rgi, nil
+	}
+
+	recordReader, err := reader.GetRecordReader(ctx, columnIndices, rowGroupIndices)
+	if err != nil {
+		return nil, fmt.Errorf("%w", err)
+	}
+
+	rgi := &Iterator{
+		ctx:    ctx,
+		reader: reader,
+		// TODO maybe set keys to selectedColumns so we include the index key in the result
+		// used to filter in Next below, see note there
+		keys:   keys,
+		config: config,
+
+		needPage:        needPage,
+		rowGroupIndices: rowGroupIndices,
+
+		recordReader: recordReader,
+		current:      &NoopRowIterator{},
+	}
+
+	return rgi, nil
+}
+
+func (r *Iterator) Next() (bool, error) {
+	next, err := r.current.Next()
+	if next || err != nil {
+		return next, err
+	}
+
+	if r.recordReader != nil && r.recordReader.Next() {
+		// NOTE: in NewRowIterator, we should maybe be setting keys to selectedColumns so we include the index key in the result
```

Comment:
Can you resolve the TODOs and NOTEs in the PR? I'm guessing these got copied over from wandb/core?